x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/notifications/AbstractAuditMessage.java

```diff
@@ -154,6 +157,7 @@ public boolean equals(Object obj) {
     /**
      * @return resource id field name used when storing a new message
      */
+    @Nullable
     protected abstract String getResourceField();
```

Comment:
There are many things that assume this is not nullable 
```
if (resourceId != null) {
            builder.field(getResourceField(), resourceId);
        }
```
In the builder for one. 

Could you change the abstraction here to be `Optional<String>` to ensure we handle all possible NPEs?